Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -12,311 +12,811 @@
+  /// an optional additional line that offers specific highlights, both based on the provided diagnostics.
+  ///
+  /// - Parameters:
+  ///   - annotatedLine: An `AnnotatedSourceLine` object containing the source line and its related diagnostics.
+  ///   - tree: A syntax tree that represents the source code file.
+  ///   - sourceLocationConverter: A `SourceLocationConverter` object used for converting between binary offset and line/column.
+  ///
+  /// - Returns: A tuple containing:
+  ///   - `highlightedSourceCode`: A string that includes the decorated source line.
+  ///   - `additionalHighlightedLine`: An optional string that provides further contextual highlights. This is `nil` if there are no additional highlight line.
+  private func decorateSourceLineWithHighlights(
+    annotatedLine: AnnotatedSourceLine,
+    inSyntaxTree tree: SyntaxProtocol,
+    using sourceLocationConverter: SourceLocationConverter
+  ) -> (highlightedSourceCode: String, additionalHighlightedLine: String?) {
+    if annotatedLine.diagnostics.isEmpty {
+      return (highlightedSourceCode: annotatedLine.sourceString, additionalHighlightedLine: nil)
     }
 
-    // Accumulate the fully annotated source files here.
-    var annotatedSource = ""
+    let highlightRanges = computeHighlightRanges(
+      forLine: annotatedLine,
+      fromTree: tree,
+      usingSourceLocationConverter: sourceLocationConverter
+    )
 
-    /// Keep track if a line missing char should be printed
-    var hasLineBeenSkipped = false
+    // Map the column ranges into index ranges within the source string itself.
+    let highlightIndexRanges: [Range<String.Index>] =
+      highlightRanges
+      .map { highlightRange in
+        let sourceStringUTF8 = annotatedLine.sourceString.utf8
+        let startIndex = sourceStringUTF8.index(sourceStringUTF8.startIndex, offsetBy: highlightRange.lowerBound - 1)
+        let endIndex = sourceStringUTF8.index(startIndex, offsetBy: highlightRange.count)
+        return startIndex..<endIndex
+      }
 
-    let maxNumberOfDigits = String(annotatedSourceLines.count).count
+    let sourceString = annotatedLine.sourceString
 
-    for (lineIndex, annotatedLine) in annotatedSourceLines.enumerated() {
-      let lineNumber = lineIndex + 1
-      guard
-        rangesToPrint.contains(where: { range in
-          range.contains(lineNumber)
-        })
-      else {
-        hasLineBeenSkipped = true
-        continue
+    var highlightedSourceCode = ""
+    var additionalHighlightedLine = ""
+    var lastIndex = sourceString.startIndex
+
+    // Iterate over each range and decorate it.
+    for highlightIndexRange in highlightIndexRanges {
+      // Append the non-highlighted part to the result string.
+      let nonHighlighted = sourceString[lastIndex..<highlightIndexRange.lowerBound]
+      highlightedSourceCode += nonHighlighted
+      additionalHighlightedLine += Array(repeating: " ", count: nonHighlighted.count).joined()
```

Comment:
No need to use `Array` here

```suggestion
      additionalHighlightedLine += String(repeating: " ", count: nonHighlighted.count)
```

Reply:
Haha, I don't know why I've used an array in the first place. Thanks for pointing out. Done.